lightning/src/ln/channelmanager.rs

```diff
@@ -3100,41 +3127,64 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 											// Phantom payments are only PendingHTLCRouting::Receive.
 											phantom_shared_secret: None,
 										});
-										match chan.get_mut().send_htlc(amt_to_forward, payment_hash, outgoing_cltv_value, htlc_source.clone(), onion_packet, &self.logger) {
-											Err(e) => {
-												if let ChannelError::Ignore(msg) = e {
-													log_trace!(self.logger, "Failed to forward HTLC with payment_hash {}: {}", log_bytes!(payment_hash.0), msg);
-												} else {
-													panic!("Stated return value requirements in send_htlc() were not met");
-												}
-												let (failure_code, data) = self.get_htlc_temp_fail_err_and_data(0x1000|7, short_chan_id, chan.get());
-												failed_forwards.push((htlc_source, payment_hash,
-													HTLCFailReason::Reason { failure_code, data }
-												));
-												continue;
-											},
-											Ok(update_add) => {
-												match update_add {
-													Some(msg) => { add_htlc_msgs.push(msg); },
-													None => {
-														// Nothing to do here...we're waiting on a remote
-														// revoke_and_ack before we can add anymore HTLCs. The Channel
-														// will automatically handle building the update_add_htlc and
-														// commitment_signed messages when we can.
-														// TODO: Do some kind of timer to set the channel as !is_live()
-														// as we don't really want others relying on us relaying through
-														// this channel currently :/.
+
+										// Attempt to forward the HTLC over all available channels
+										// to the peer, but attempt to forward the HTLC over the
+										// channel specified in the onion payload first.
+										let mut counterparty_channel_ids = peer_state.channel_by_id.keys()
```

Comment:
If you can think of a more efficient way of doing this, please me know :).

Reply:
I think using `iter::once()` and `iter::chain()` should allow you to get rid of the insert. E.g.:
```Rust
let mut counterparty_channel_ids = core::iter::once(forward_chan_id)
    .chain(peer_state.channel_by_id.keys()
                .filter(|&chan_id| *chan_id != forward_chan_id)
                .map(|&chan_id| chan_id))
    .collect::<Vec<_>>();
```

Reply:
Nice thanks! 13dbfe3